### DISCUSSION TITLE
Send/Sync Bus Devices

### DISCUSSION THREAD

**Discussion by posk-io** [cloud-hypervisor/cloud-hypervisor#4854] [Ideas]
At the moment bus devices in Cloud Hypervisor live under mutexes:

https://github.com/cloud-hypervisor/cloud-hypervisor/blob/6722c303b05c0b3b7f751ec43927db315d4cd52c/vm-device/src/bus.rs#L98

This results in locking the device mutex on every IO: 

https://github.com/cloud-hypervisor/cloud-hypervisor/blob/6722c303b05c0b3b7f751ec43927db315d4cd52c/vm-device/src/bus.rs#L236

This is not desirable for devices that are designed to operate concurrently per-vcpu. For example, CrosVM allows "sync" devices that don't require mutexes "on the bus":

https://github.com/google/crosvm/blob/24ed80fd99dd6e7826834ace2b0cce215c65e80e/devices/src/bus.rs#L288

Would it be possible for Cloud Hypervisor to support such devices? Would you accept a PR to this end?

The use case is to use PIO with shared mem to emulate a synchronous hypercall, as suggested here: https://lore.kernel.org/lkml/b1b04160-1604-8281-4c82-09b1f84ba86c@redhat.com/

**Comment by rbradford** [comment]
You could have a per vCPU MMIO/PCI bus as the exits are handled per vCPU.

**Reply by rbradford** [reply]
Further, i'm not sure we have any devices that would take advantage of that right now?

**Reply by posk-io** [reply]
MMIO/PCI devices seem to be much more complicated to implement; a PIO "device" is just an asm line in the guest to poke a port address, and a two-fn trait implementation in the host VMM.

Re: "existing devices": the proposed change could be used in experimental/research/prototyping set-ups, where the relative simplicity of implementing a "PIO device" vs an MMIO/PCI device is of even more value.

**Reply by rbradford** [reply]
My concern is that it would be a lot of work in Cloud Hypervisor for something that doesn't yet have a use?
